### PR TITLE
Add env configuration support

### PR DESCRIPTION
**Pull request by shadowhand**
Often projects will need some kind of configuration for the application
environment. The current best practice is to store configuration in an
ini file that is parsed.

### CODE REVIEW

File: tests/Configuration/EnvConfigurationTest.php [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+    {
+        return [
+            new EnvConfiguration(__DIR__ . '/../_fixture/.env'),
+        ];
+    }
+
+    public function testApply()
+    {
+        if (!class_exists('josegonzalez\Dotenv\Loader')) {
+            $this->markTestSkipped('Plates is not installed');
```
Comment (elazar) [inline tests/Configuration/EnvConfigurationTest.php]: Plates? :wink:

File: src/Exception/EnvException.php [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+{
+    /**
+     * @param string $path
+     *
+     * @return static
+     */
+    public static function invalidFile($path)
+    {
+        return new static(sprintf(
+            'Environment file `%s` does not exist or is not readable',
```
Comment (shadowhand) [inline src/Exception/EnvException.php]: @elazar fixed. ;)

File: src/Configuration/EnvConfiguration.php [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+            return $env->withData($values);
+        });
+    }
+
+    private function getProjectRoot()
+    {
+        // Assuming that __DIR__ is vendor/sparkphp/spark/src/Configuration ...
+        // we should be able to remove these last 5 segments from the current
+        // directory to get the root path of the application.
+        return dirname(dirname(dirname(dirname(dirname(__DIR__))))) . DIRECTORY_SEPARATOR;
```
Comment (elazar) [inline src/Configuration/EnvConfiguration.php]: There a reason we can't just have this check for both the dependency and local clone cases, with preference given to the former?

Comment (shadowhand) [inline src/Configuration/EnvConfiguration.php]: Sounds good, I'll do that.

File: docs/index.md [hunk 4]
```diff
@@ -113,8 +113,23 @@ The following configurations are typically used by default:
 
 The following configurations are available but not used by default:
 
+* [`EnvConfiguration`](https://github.com/sparkphp/spark/blob/master/src/Configuration/EnvConfiguration.php) - Use [Dotenv](https://github.com/josegonzalez/php-dotenv) to populate the content of [`Env`](https://github.com/sparkphp/spark/blob/master/src/Env.php)
 * [`PlatesResponderConfiguration`](https://github.com/sparkphp/spark/blob/master/src/Configuration/PlatesResponderConfiguration.php) - Use [Plates](http://platesphp.com/) as the default [responder](#responders)
 
+### Env Configuration
```
Comment (shadowhand) [inline docs/index.md]: @elazar now with docs!

Comment (elazar) [inline docs/index.md]: :trophy:

File: src/Configuration/EnvConfiguration.php [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+    private function detectEnvFile()
+    {
+        $env = DIRECTORY_SEPARATOR . '.env';
+        $dir = dirname(dirname(__DIR__));
+
+        do {
+            if (is_file($dir . $env)) {
+                return $dir . $env;
+            }
+        } while ($dir = dirname($dir) && is_readable($dir));
```
Comment (shadowhand) [inline src/Configuration/EnvConfiguration.php]: @elazar this was my solution for handling file detection. Seemed easier than doing multiple methods to check for files.

Comment (elazar) [inline src/Configuration/EnvConfiguration.php]: :+1:

**Pull request merged**